src/Data/Formatter/DateTime.purs

```diff
@@ -0,0 +1,426 @@
+    , (PC.try $ PS.string "a") *> map Meridiem cb
+    , (PC.try $ PS.string "mm") *> map Minutes cb
+    , (PC.try $ PS.string "ss") *> map Seconds cb
+    , (PC.try $ PS.string "SSS") *> map Milliseconds cb
+    , (Placeholder <$> placeholderContent <*> cb)
+    , (PS.eof $> End)
+    ]
+
+formatParser ∷ P.Parser String Formatter
+formatParser =
+  Lazy.fix \f → map roll $ formatterFParser f
+
+formatF
+  ∷ ∀ a
+  . (a → String)
+  → DT.DateTime
+  → FormatterF a
+  → String
+formatF cb dt@(DT.DateTime d t) = case _ of
+  YearFull a →
+    (show $ fromEnum $ D.year d) <> cb a
+  YearTwoDigits a →
+    show ((fromEnum $ D.year d) - 1900) <> cb a
+  YearAbsolute a →
+    show (fromEnum $ D.year d) <> cb a
+  MonthFull a →
+    show (D.month d) <> cb a
+  MonthShort a →
+    printShortMonth (D.month d) <> cb a
+  MonthTwoDigits a →
+    show (fromEnum $ D.month d) <> cb a
+  DayOfMonth a →
+    show (fromEnum $ D.day d) <> cb a
+  UnixTimestamp a →
+    (show $ Int.floor $ (_ / 1000.0) $ Dur.unMilliseconds $ unInstant $ fromDateTime dt) <> cb a
+  DayOfWeek a →
+    show (fromEnum $ D.weekday d) <> cb a
+  Hours24 a →
+    show (fromEnum $ T.hour t) <> cb a
+  Hours12 a →
+    show ((fromEnum $ T.hour t) `mod` 12) <> cb a
+  Meridiem a →
+    (if (fromEnum $ T.hour t) > 12 then "PM" else "AM") <> cb a
+  Minutes a →
+    show (fromEnum $ T.minute t) <> cb a
+  Seconds a →
+    show (fromEnum $ T.second t) <> cb a
+  Milliseconds a →
+    show (fromEnum $ T.millisecond t) <> cb a
+  Placeholder s a →
+    s <> cb a
+  End → ""
+
+
+format ∷ Formatter → DT.DateTime → String
+format f dt = formatF (flip format dt) dt $ unroll f
+
+formatDateTime ∷ String → DT.DateTime → Either String String
+formatDateTime pattern datetime =
+  parseFormatString pattern <#> flip format datetime
+
+unformat ∷ Formatter → String → Either String DT.DateTime
+unformat f s =
+  let
+    run =
+      runState
+        (P.runParserT (P.PState {input: s, position: initialPos}) $ unformatParser f)
+        initialAccum
+  in
+    case run of
+      Tuple (Left (P.ParseError {message})) state → Left message
+      Tuple _ accum → unformatAccumToDateTime accum
+
+data Meridiem = AM | PM
+
+derive instance eqMeridiem ∷ Eq Meridiem
+
+-- TODO: consider using Map Int
+type UnformatAccum =
+  { year ∷ Maybe Int
+  , month ∷ Maybe Int
+  , day ∷ Maybe Int
+  , hour ∷ Maybe Int
+  , minute ∷ Maybe Int
+  , second ∷ Maybe Int
+  , millisecond ∷ Maybe Int
+  , dayOfWeek ∷ Maybe Int
+  , meridiem ∷ Maybe Meridiem
+  }
+
+initialAccum ∷ UnformatAccum
+initialAccum =
+  { year: Nothing
+  , month: Nothing
+  , day: Nothing
+  , hour: Nothing
+  , minute: Nothing
+  , second: Nothing
+  , millisecond: Nothing
+  , dayOfWeek: Nothing
+  , meridiem: Nothing
+  }
+
+unformatAccumToDateTime ∷ UnformatAccum → Either String DT.DateTime
+unformatAccumToDateTime a =
+  DT.DateTime
+    <$> (D.canonicalDate
+           <$> (maybe (Left "incorrect year") pure $ toEnum $ fromMaybe zero a.year)

```

Comment:
The capitalisation of `Incorrect` here differs from some of the cases below.

Perhaps we could phrase it a little nicer too? "Could not parse year", etc. as "incorrect" could imply the value is somehow wrong (which is unknowable) rather than it just not matching the expected format.